File: week1_1.5/WeightedQU.cpp
#include "WeightedQU.h"

WeightedQU::WeightedQU(int size){
    id.resize(size);
    sz.resize(size);
    for (int i=0;i<size;i++){
        id[i] = i;
        sz[i] = 1;
    }
}

void WeightedQU::Union(int a, int b){
    int ra = root(a), rb = root(b);
    if (sz[ra] >= sz[rb]){
        //append tree rb to ra
        id[rb] = ra;
        //update sz[ra]
        sz[ra] += sz[rb];
    }else{
        //append tree ra to rb
        id[ra] = rb;
        //update sz[rb]
        sz[rb] += sz[ra];
    }
}

bool WeightedQU::Connected(int a, int b){
    return root(a) == root(b);
}

int WeightedQU::root(int a){
    while(id[a] != a) a = id[a];
    return a;
}

File: src/chapter1_5/WeightedQU.java
package ch1.DynamicConnection;

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.Stopwatch;

public class WeightedQU {
    private int[] id;
    private int[] sz;

    public WeightedQU(int n) {
        id = new int[n];
        sz = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
            sz[i] = 0;
        }
    }

    private int find(int a) {
        while (a != id[a]) {
            a = id[a];
        }
        return a;
    }

    public boolean connected(int a, int b) {
        return find(a) == find(b);
    }

    public void union(int a, int b) {
        int ida = find(a);
        int idb = find(b);
        if (ida == idb) return;
        if (sz[ida] > sz[idb]) {
            id[idb] = ida;
            sz[ida] += sz[idb];
        } else {
            id[ida] = idb;
            sz[idb] += sz[ida];
        }
    }

    public static void main(String[] args) {
        In in = new In(args[0]);
        int n = in.readInt();
        WeightedQU uf = new WeightedQU(n);
        Stopwatch watch = new Stopwatch();
        for (int i = 0; i < n; i++) {
            int a = in.readInt();
            int b = in.readInt();
            if (uf.connected(a, b)) {
                ;//StdOut.println("(" + a + ", " + b + ") is connected.");
            } else {
                uf.union(a, b);
            }
        }
        StdOut.println(n + " data: " + watch.elapsedTime() + "s.");
    }
}
